ecc_enclave/enclave/shim.h

```diff
@@ -6,27 +6,130 @@
 
 #pragma once
 
+#include <stdbool.h>
 #include <map>
-#include <set>
 #include <string>
 #include <vector>
 
-typedef std::map<std::string, std::string> write_set_t;
-typedef std::set<std::string> read_set_t;
-typedef std::map<void*, std::pair<read_set_t*, write_set_t*>> context_t;
+typedef void* fpc_ctx_t;
 
-// shim put/get
-void get_state(const char* key, uint8_t* val, uint32_t max_val_len, uint32_t* val_len, void* ctx);
-void put_state(const char* key, uint8_t* val, uint32_t val_len, void* ctx);
+// TODO: Some Meta-questions
+// - should we prefix functions with fpc_ or alike to make them more distinct?
+//   Alternatively, we might go C++ and define a shim object like the go shim
+//   (but then maybe we do not necessarily want to force folks to go to C++ and
+//    still enable C?)
+
+// Function which FPC chaincode has to implement
+// ==================================================
+int invoke(const char* args,
```

Comment:
why moving this from chaincode.h to here?

Reply:
aeh, because i didn't even know of the existance of chaincode.h: none of the examples was using it ...
Seems, though, that only invoke but not inovke_enc would be exposed?  If so, why not move invoke to shim.h and invoke_enc to shim_internals.h?  Do we need two header files exposed to the application?

Reply:
invoke_enc doesn't even really need a function prototype, in particular not in a separate header file, as it is used and defined in same file, so i just removed chaincode.h ..